models/helper.go

```diff
@@ -25,6 +25,25 @@ func QueryFilterByConceptIdsHelper(query *gorm.DB, sourceId int, filterConceptId
 	return query
 }
 
+// Same as Query Filter above but adds additional value filter as well
+func QueryFilterByConceptIdsAndValuesHelper(query *gorm.DB, sourceId int, filterConceptIdsAndValues []utils.CustomConceptVariableDef,
+	omopDataSource *utils.DbAndSchema, resultSchemaName string, personIdFieldForObservationJoin string) *gorm.DB {
+	// iterate over the filterConceptIds, adding a new INNER JOIN and filters for each, so that the resulting set is the
+	// set of persons that have a non-null value for each and every one of the concepts:
+	for i, filterConceptIdAndValue := range filterConceptIdsAndValues {
+		observationTableAlias := fmt.Sprintf("observation_filter_%d", i)
+		log.Printf("Adding extra INNER JOIN with alias %s", observationTableAlias)
+		query = query.Joins("INNER JOIN "+omopDataSource.Schema+".observation_continuous as "+observationTableAlias+omopDataSource.GetViewDirective()+" ON "+observationTableAlias+".person_id = "+personIdFieldForObservationJoin).
+			Where(observationTableAlias+".observation_concept_id = ?", filterConceptIdAndValue.ConceptId).
+			Where(GetConceptValueNotNullCheckBasedOnConceptType(observationTableAlias, sourceId, filterConceptIdAndValue.ConceptId))
```

Comment:
you could move this to else after the `if len(filterConceptIdAndValue.ConceptValues) > 0` below

Reply:
Hmm I don't think this code belongs to an else block after the check. The reason is we still need to limit the observation to this particular conceptid. If the value filters are passed in we then AND the clause for filter by value. It's not an either/or relationship.

If the value_as_concept_id is shared with other concepts we could potentially pick up with unwanted rows relating other concept ids.

Reply:
I meant: no need to check for "not null" if we are already checking for a specific value.

Reply:
yes I see what you mean now, updated